File: init/graphics.c
#include "graphics.h"
#include "font8x8.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <sys/ioctl.h>
#include <sys/mman.h>
#include <linux/fb.h>
#include <linux/kd.h>

/* Internal State */
static int s_fb_fd = -1;
static uint8_t *s_fb_mem = NULL;
static uint8_t *s_backbuffer = NULL;
static uint8_t *s_draw_buffer = NULL;
static struct fb_fix_screeninfo s_finfo;
static struct fb_var_screeninfo s_vinfo;
static uint32_t s_bytes_per_pixel = 0;
static bool s_initialized = false;

static inline uint32_t pack_color(struct color c)
{
    uint32_t r = (c.r >> (8 - s_vinfo.red.length)) << s_vinfo.red.offset;
    uint32_t g = (c.g >> (8 - s_vinfo.green.length)) << s_vinfo.green.offset;
    uint32_t b = (c.b >> (8 - s_vinfo.blue.length)) << s_vinfo.blue.offset;
    uint32_t a = (s_vinfo.transp.length > 0) ?
                 ((c.a >> (8 - s_vinfo.transp.length)) << s_vinfo.transp.offset) : 0;
    return r | g | b | a;
}

static inline struct color unpack_color(uint32_t p)
{
    struct color c;
    c.r = (uint8_t)(((p >> s_vinfo.red.offset) & ((1 << s_vinfo.red.length) - 1)) << (8 - s_vinfo.red.length));
    c.g = (uint8_t)(((p >> s_vinfo.green.offset) & ((1 << s_vinfo.green.length) - 1)) << (8 - s_vinfo.green.length));
    c.b = (uint8_t)(((p >> s_vinfo.blue.offset) & ((1 << s_vinfo.blue.length) - 1)) << (8 - s_vinfo.blue.length));
    c.a = (s_vinfo.transp.length > 0) ?
          (uint8_t)(((p >> s_vinfo.transp.offset) & ((1 << s_vinfo.transp.length) - 1)) << (8 - s_vinfo.transp.length)) : 255;
    return c;
}

void graphics_silence_vt(bool enable)
{
    int tty_fd = open("/dev/tty0", O_RDWR);
    if (tty_fd < 0) tty_fd = open("/dev/tty1", O_RDWR);
    if (tty_fd < 0) tty_fd = open("/dev/console", O_RDWR);
    if (tty_fd >= 0) {
        ioctl(tty_fd, KDSETMODE, enable ? KD_GRAPHICS : KD_TEXT);
        close(tty_fd);
    }
}

int graphics_init(void)
{
    if (s_initialized)
        return 0;

    const char *fb_path = "/dev/fb0";
    s_fb_fd = open(fb_path, O_RDWR);
    if (s_fb_fd < 0) {
        perror("graphics_init: open /dev/fb0 failed");
        return -1;
    }

    if (ioctl(s_fb_fd, FBIOGET_FSCREENINFO, &s_finfo) < 0) {
        perror("graphics_init: ioctl FBIOGET_FSCREENINFO failed");
        close(s_fb_fd);
        s_fb_fd = -1;
        return -1;
    }

    if (ioctl(s_fb_fd, FBIOGET_VSCREENINFO, &s_vinfo) < 0) {
        perror("graphics_init: ioctl FBIOGET_VSCREENINFO failed");
        close(s_fb_fd);
        s_fb_fd = -1;
        return -1;
    }

    s_bytes_per_pixel = s_vinfo.bits_per_pixel / 8;
    if (s_bytes_per_pixel != 4 && s_bytes_per_pixel != 2 && s_bytes_per_pixel != 3) {
        fprintf(stderr, "graphics_init: unsupported bits per pixel: %u\n", s_vinfo.bits_per_pixel);
        close(s_fb_fd);
        s_fb_fd = -1;
        return -1;
    }

    s_fb_mem = (uint8_t *)mmap(NULL, s_finfo.smem_len, PROT_READ | PROT_WRITE, MAP_SHARED, s_fb_fd, 0);
    if (s_fb_mem == MAP_FAILED) {
        perror("graphics_init: mmap failed");
        close(s_fb_fd);
        s_fb_fd = -1;
        s_fb_mem = NULL;
        return -1;
    }

    /* Allocate RAM back-buffer for tear-free 60 FPS double-buffering */
    s_backbuffer = (uint8_t *)malloc(s_finfo.smem_len);
    if (s_backbuffer) {
        s_draw_buffer = s_backbuffer;
        memset(s_backbuffer, 0, s_finfo.smem_len);
    } else {
        /* Fall back to direct framebuffer if RAM is tight */
        s_draw_buffer = s_fb_mem;
    }

    /* Silence Linux VT cursor and console text while in graphics */
    graphics_silence_vt(true);

    s_initialized = true;
    return 0;
}

void graphics_shutdown(void)
{
    if (!s_initialized)
        return;

    /* Restore Linux VT text mode so TTY console is normal */
    graphics_silence_vt(false);

    if (s_backbuffer) {
        free(s_backbuffer);
        s_backbuffer = NULL;
    }
    s_draw_buffer = NULL;

    if (s_fb_mem && s_fb_mem != MAP_FAILED) {
        munmap(s_fb_mem, s_finfo.smem_len);
        s_fb_mem = NULL;
    }

    if (s_fb_fd >= 0) {
        close(s_fb_fd);
        s_fb_fd = -1;
    }

    s_initialized = false;
}

void graphics_swap_buffers(void)
{
    if (s_initialized && s_backbuffer && s_fb_mem) {
        memcpy(s_fb_mem, s_backbuffer, s_finfo.smem_len);
    }
}

void graphics_copy_to_draw_buffer(const void *src, size_t size)
{
    if (s_draw_buffer && src) {
        size_t copy_len = (size <= s_finfo.smem_len) ? size : s_finfo.smem_len;
        memcpy(s_draw_buffer, src, copy_len);
    }
}

uint8_t *graphics_get_draw_buffer(void)
{
    return s_draw_buffer;
}

uint32_t graphics_width(void)  { return s_vinfo.xres; }
uint32_t graphics_height(void) { return s_vinfo.yres; }
uint32_t graphics_pitch(void)  { return s_finfo.line_length; }
uint32_t graphics_bpp(void)    { return s_vinfo.bits_per_pixel; }

void graphics_put_pixel(int x, int y, struct color c)
{
    if (!s_initialized || !s_draw_buffer)
        return;

    /* Strict bounds check */
    if (x < 0 || (uint32_t)x >= s_vinfo.xres || y < 0 || (uint32_t)y >= s_vinfo.yres)
        return;

    size_t offset = (size_t)y * s_finfo.line_length + (size_t)x * s_bytes_per_pixel;
    if (offset + s_bytes_per_pixel > s_finfo.smem_len)
        return;

    uint32_t pval = pack_color(c);

    if (s_bytes_per_pixel == 4) {
        *(volatile uint32_t *)(s_draw_buffer + offset) = pval;
    } else if (s_bytes_per_pixel == 2) {
        *(volatile uint16_t *)(s_draw_buffer + offset) = (uint16_t)pval;
    } else if (s_bytes_per_pixel == 3) {
        s_draw_buffer[offset + 0] = c.b;
        s_draw_buffer[offset + 1] = c.g;
        s_draw_buffer[offset + 2] = c.r;
    }
}

struct color graphics_get_pixel(int x, int y)
{
    if (!s_initialized || !s_draw_buffer)
        return COLOR_BLACK;

    if (x < 0 || (uint32_t)x >= s_vinfo.xres || y < 0 || (uint32_t)y >= s_vinfo.yres)
        return COLOR_BLACK;

    size_t offset = (size_t)y * s_finfo.line_length + (size_t)x * s_bytes_per_pixel;
    if (offset + s_bytes_per_pixel > s_finfo.smem_len)
        return COLOR_BLACK;

    uint32_t pval = 0;
    if (s_bytes_per_pixel == 4) {
        pval = *(volatile uint32_t *)(s_draw_buffer + offset);
    } else if (s_bytes_per_pixel == 2) {
        pval = *(volatile uint16_t *)(s_draw_buffer + offset);
    } else if (s_bytes_per_pixel == 3) {
        struct color c;
        c.b = s_draw_buffer[offset + 0];
        c.g = s_draw_buffer[offset + 1];
        c.r = s_draw_buffer[offset + 2];
        c.a = 255;
        return c;
    }

    return unpack_color(pval);
}

void graphics_clear(struct color c)
{
    if (!s_initialized || !s_draw_buffer)
        return;

    uint32_t pval = pack_color(c);

    for (uint32_t y = 0; y < s_vinfo.yres; y++) {
        uint8_t *row = s_draw_buffer + (size_t)y * s_finfo.line_length;
        if (s_bytes_per_pixel == 4) {
            uint32_t *dst = (uint32_t *)row;
            for (uint32_t x = 0; x < s_vinfo.xres; x++) {
                dst[x] = pval;
            }
        } else if (s_bytes_per_pixel == 2) {
            uint16_t *dst = (uint16_t *)row;
            for (uint32_t x = 0; x < s_vinfo.xres; x++) {
                dst[x] = (uint16_t)pval;
            }
        } else {
            for (uint32_t x = 0; x < s_vinfo.xres; x++) {
                graphics_put_pixel((int)x, (int)y, c);
            }
        }
    }
}

void graphics_fill_rect(int x, int y, int w, int h, struct color c)
{
    if (!s_initialized || !s_draw_buffer || w <= 0 || h <= 0)
        return;

    int x0 = x;
    int y0 = y;
    int x1 = x + w;
    int y1 = y + h;

    /* Clip rectangle */
    if (x0 < 0) x0 = 0;
    if (y0 < 0) y0 = 0;
    if (x1 > (int)s_vinfo.xres) x1 = (int)s_vinfo.xres;
    if (y1 > (int)s_vinfo.yres) y1 = (int)s_vinfo.yres;

    if (x0 >= x1 || y0 >= y1)
        return;

    uint32_t pval = pack_color(c);

    for (int cy = y0; cy < y1; cy++) {
        uint8_t *row = s_draw_buffer + (size_t)cy * s_finfo.line_length;
        if (s_bytes_per_pixel == 4) {
            uint32_t *dst = (uint32_t *)row;
            for (int cx = x0; cx < x1; cx++) {
                dst[cx] = pval;
            }
        } else if (s_bytes_per_pixel == 2) {
            uint16_t *dst = (uint16_t *)row;
            for (int cx = x0; cx < x1; cx++) {
                dst[cx] = (uint16_t)pval;
            }
        } else {
            for (int cx = x0; cx < x1; cx++) {
                graphics_put_pixel(cx, cy, c);
            }
        }
    }
}

void graphics_draw_rect(int x, int y, int w, int h, struct color c)
{
    if (w <= 0 || h <= 0)
        return;

    graphics_draw_line(x, y, x + w - 1, y, c);
    graphics_draw_line(x, y + h - 1, x + w - 1, y + h - 1, c);
    graphics_draw_line(x, y, x, y + h - 1, c);
    graphics_draw_line(x + w - 1, y, x + w - 1, y + h - 1, c);
}

void graphics_draw_line(int x0, int y0, int x1, int y1, struct color c)
{
    int dx = abs(x1 - x0);
    int sx = (x0 < x1) ? 1 : -1;
    int dy = -abs(y1 - y0);
    int sy = (y0 < y1) ? 1 : -1;
    int err = dx + dy;

    while (1) {
        graphics_put_pixel(x0, y0, c);
        if (x0 == x1 && y0 == y1)
            break;
        int e2 = 2 * err;
        if (e2 >= dy) {
            err += dy;
            x0 += sx;
        }
        if (e2 <= dx) {
            err += dx;
            y0 += sy;
        }
    }
}

void graphics_draw_circle(int cx, int cy, int radius, struct color c)
{
    if (radius < 0)
        return;

    int x = radius;
    int y = 0;
    int err = 0;

    while (x >= y) {
        graphics_put_pixel(cx + x, cy + y, c);
        graphics_put_pixel(cx + y, cy + x, c);
        graphics_put_pixel(cx - y, cy + x, c);
        graphics_put_pixel(cx - x, cy + y, c);
        graphics_put_pixel(cx - x, cy - y, c);
        graphics_put_pixel(cx - y, cy - x, c);
        graphics_put_pixel(cx + y, cy - x, c);
        graphics_put_pixel(cx + x, cy - y, c);

        if (err <= 0) {
            y += 1;
            err += 2 * y + 1;
        }
        if (err > 0) {
            x -= 1;
            err -= 2 * x + 1;
        }
    }
}

void graphics_fill_circle(int cx, int cy, int radius, struct color c)
{
    if (radius < 0)
        return;

    int r2 = radius * radius;
    for (int dy = -radius; dy <= radius; dy++) {
        int dx_max = 0;
        while ((dx_max + 1) * (dx_max + 1) + dy * dy <= r2) {
            dx_max++;
        }
        graphics_draw_line(cx - dx_max, cy + dy, cx + dx_max, cy + dy, c);
    }
}

void graphics_draw_char(int x, int y, char ch, struct color fg, struct color bg)
{
    unsigned char uch = (unsigned char)ch;
    if (uch < 32 || uch > 127)
        uch = '?';

    const uint8_t *bitmap = font8x8_basic[(int)uch - 32];

    for (int row = 0; row < 8; row++) {
        uint8_t byte = bitmap[row];
        for (int col = 0; col < 8; col++) {
            if (byte & (0x80 >> col)) {
                graphics_put_pixel(x + col, y + row, fg);
            } else if (bg.a > 0) {
                graphics_put_pixel(x + col, y + row, bg);
            }
        }
    }
}

void graphics_draw_text(int x, int y, const char *str, struct color fg, struct color bg)
{
    if (!str)
        return;

    int cur_x = x;
    int cur_y = y;

    while (*str) {
        if (*str == '\n') {
            cur_x = x;
            cur_y += 10;
        } else if (*str == '\t') {
            cur_x += 32;
        } else {
            graphics_draw_char(cur_x, cur_y, *str, fg, bg);
            cur_x += 8;
        }
        str++;
    }
}

File: ui/src/graphics.rs
use crate::color::Color;
use crate::font::get_glyph;

use std::fs::{File, OpenOptions};
use std::os::unix::io::{AsRawFd, RawFd};
use std::ptr;

const FBIOGET_FSCREENINFO: libc::c_int = 0x4602;
const FBIOGET_VSCREENINFO: libc::c_int = 0x4600;
const KDSETMODE: libc::c_int = 0x4B3A;
const KD_TEXT: libc::c_int = 0;
const KD_GRAPHICS: libc::c_int = 1;

#[repr(C)]
struct FbBitfield {
    offset: u32,
    length: u32,
    msb_right: u32,
}

#[repr(C)]
struct FbVarScreeninfo {
    xres: u32,
    yres: u32,
    xres_virtual: u32,
    yres_virtual: u32,
    xoffset: u32,
    yoffset: u32,
    bits_per_pixel: u32,
    grayscale: u32,
    red: FbBitfield,
    green: FbBitfield,
    blue: FbBitfield,
    transp: FbBitfield,
    nonstd: u32,
    activate: u32,
    height: u32,
    width: u32,
    accel_flags: u32,
    pixclock: u32,
    left_margin: u32,
    right_margin: u32,
    upper_margin: u32,
    lower_margin: u32,
    hsync_len: u32,
    vsync_len: u32,
    sync: u32,
    vmode: u32,
    rotate: u32,
    colorspace: u32,
    reserved: [u32; 4],
}

#[repr(C)]
struct FbFixScreeninfo {
    id: [u8; 16],
    smem_start: libc::c_ulong,
    smem_len: u32,
    fb_type: u32,
    type_aux: u32,
    visual: u32,
    xpanstep: u16,
    ypanstep: u16,
    ywrapstep: u16,
    line_length: u32,
    mmio_start: libc::c_ulong,
    mmio_len: u32,
    accel: u32,
    capabilities: u16,
    reserved: [u16; 2],
}

#[allow(dead_code)]
pub struct Graphics {
    fb_file: File,
    mmap_ptr: *mut u8,
    smem_len: usize,
    width: u32,
    height: u32,
    pitch: u32,
    bpp: u32,
    bytes_per_pixel: u32,
    red: FbBitfield,
    green: FbBitfield,
    blue: FbBitfield,
    transp: FbBitfield,
    backbuffer: Vec<u8>,
    tty_fd: Option<RawFd>,
}

#[allow(dead_code)]
impl Graphics {
    pub fn new() -> Result<Self, String> {
        let fb_file = OpenOptions::new()
            .read(true)
            .write(true)
            .open("/dev/fb0")
            .map_err(|e| format!("Failed to open /dev/fb0: {}", e))?;

        let raw_fd = fb_file.as_raw_fd();

        let mut finfo: FbFixScreeninfo = unsafe { std::mem::zeroed() };
        if unsafe { libc::ioctl(raw_fd, FBIOGET_FSCREENINFO, &mut finfo) } < 0 {
            return Err("ioctl FBIOGET_FSCREENINFO failed".to_string());
        }

        let mut vinfo: FbVarScreeninfo = unsafe { std::mem::zeroed() };
        if unsafe { libc::ioctl(raw_fd, FBIOGET_VSCREENINFO, &mut vinfo) } < 0 {
            return Err("ioctl FBIOGET_VSCREENINFO failed".to_string());
        }

        let bpp = vinfo.bits_per_pixel;
        let bytes_per_pixel = bpp / 8;
        if bytes_per_pixel != 4 && bytes_per_pixel != 2 && bytes_per_pixel != 3 {
            return Err(format!("Unsupported bits per pixel: {}", bpp));
        }

        let smem_len = finfo.smem_len as usize;
        let mmap_ptr = unsafe {
            libc::mmap(
                ptr::null_mut(),
                smem_len,
                libc::PROT_READ | libc::PROT_WRITE,
                libc::MAP_SHARED,
                raw_fd,
                0,
            )
        };

        if mmap_ptr == libc::MAP_FAILED {
            return Err("mmap /dev/fb0 failed".to_string());
        }

        let backbuffer = vec![0u8; smem_len];

        // Silence Linux Virtual Terminal cursor & text output
        let mut tty_fd = None;
        for path in &["/dev/tty0", "/dev/tty1", "/dev/console"] {
            let fd = unsafe { libc::open(path.as_ptr() as *const libc::c_char, libc::O_RDWR) };
            if fd >= 0 {
                unsafe { libc::ioctl(fd, KDSETMODE, KD_GRAPHICS) };
                tty_fd = Some(fd);
                break;
            }
        }

        Ok(Self {
            fb_file,
            mmap_ptr: mmap_ptr as *mut u8,
            smem_len,
            width: vinfo.xres,
            height: vinfo.yres,
            pitch: finfo.line_length,
            bpp,
            bytes_per_pixel,
            red: vinfo.red,
            green: vinfo.green,
            blue: vinfo.blue,
            transp: vinfo.transp,
            backbuffer,
            tty_fd,
        })
    }

    pub fn width(&self) -> u32 { self.width }
    pub fn height(&self) -> u32 { self.height }
    pub fn pitch(&self) -> u32 { self.pitch }
    pub fn bpp(&self) -> u32 { self.bpp }
    pub fn buffer_size(&self) -> usize { self.smem_len }

    #[inline(always)]
    pub fn pack_color(&self, c: Color) -> u32 {
        let r = ((c.r >> (8 - self.red.length)) as u32) << self.red.offset;
        let g = ((c.g >> (8 - self.green.length)) as u32) << self.green.offset;
        let b = ((c.b >> (8 - self.blue.length)) as u32) << self.blue.offset;
        let a = if self.transp.length > 0 {
            ((c.a >> (8 - self.transp.length)) as u32) << self.transp.offset
        } else {
            0
        };
        r | g | b | a
    }

    #[inline(always)]
    pub fn put_pixel(&mut self, x: i32, y: i32, c: Color) {
        if x < 0 || (x as u32) >= self.width || y < 0 || (y as u32) >= self.height {
            return;
        }

        let offset = (y as usize) * (self.pitch as usize) + (x as usize) * (self.bytes_per_pixel as usize);
        if offset + (self.bytes_per_pixel as usize) > self.smem_len {
            return;
        }

        let pval = self.pack_color(c);
        let bpp = self.bytes_per_pixel;

        if bpp == 4 {
            let slice = &mut self.backbuffer[offset..offset + 4];
            slice.copy_from_slice(&pval.to_ne_bytes());
        } else if bpp == 2 {
            let slice = &mut self.backbuffer[offset..offset + 2];
            slice.copy_from_slice(&(pval as u16).to_ne_bytes());
        } else if bpp == 3 {
            self.backbuffer[offset] = c.b;
            self.backbuffer[offset + 1] = c.g;
            self.backbuffer[offset + 2] = c.r;
        }
    }

    pub fn get_pixel(&self, x: i32, y: i32) -> Color {
        if x < 0 || (x as u32) >= self.width || y < 0 || (y as u32) >= self.height {
            return Color::BLACK;
        }

        let offset = (y as usize) * (self.pitch as usize) + (x as usize) * (self.bytes_per_pixel as usize);
        if offset + (self.bytes_per_pixel as usize) > self.smem_len {
            return Color::BLACK;
        }

        if self.bytes_per_pixel == 4 {
            let bytes: [u8; 4] = self.backbuffer[offset..offset + 4].try_into().unwrap_or([0; 4]);
            let pval = u32::from_ne_bytes(bytes);
            let r = (((pval >> self.red.offset) & ((1 << self.red.length) - 1)) << (8 - self.red.length)) as u8;
            let g = (((pval >> self.green.offset) & ((1 << self.green.length) - 1)) << (8 - self.green.length)) as u8;
            let b = (((pval >> self.blue.offset) & ((1 << self.blue.length) - 1)) << (8 - self.blue.length)) as u8;
            Color::rgba(r, g, b, 255)
        } else {
            Color::BLACK
        }
    }

    pub fn clear(&mut self, c: Color) {
        let pval = self.pack_color(c);
        let bytes = pval.to_ne_bytes();
        let bpp = self.bytes_per_pixel as usize;

        if bpp == 4 {
            for chunk in self.backbuffer.chunks_exact_mut(4) {
                chunk.copy_from_slice(&bytes);
            }
        } else {
            for y in 0..self.height {
                for x in 0..self.width {
                    self.put_pixel(x as i32, y as i32, c);
                }
            }
        }
    }

    pub fn fill_rect(&mut self, x: i32, y: i32, w: i32, h: i32, c: Color) {
        if w <= 0 || h <= 0 {
            return;
        }

        let x0 = x.clamp(0, self.width as i32);
        let y0 = y.clamp(0, self.height as i32);
        let x1 = (x + w).clamp(0, self.width as i32);
        let y1 = (y + h).clamp(0, self.height as i32);

        if x0 >= x1 || y0 >= y1 {
            return;
        }

        let pval = self.pack_color(c);
        let bytes = pval.to_ne_bytes();

        for cy in y0..y1 {
            let row_offset = (cy as usize) * (self.pitch as usize);
            for cx in x0..x1 {
                let offset = row_offset + (cx as usize) * (self.bytes_per_pixel as usize);
                if self.bytes_per_pixel == 4 {
                    self.backbuffer[offset..offset + 4].copy_from_slice(&bytes);
                } else {
                    self.put_pixel(cx, cy, c);
                }
            }
        }
    }

    pub fn draw_rect(&mut self, x: i32, y: i32, w: i32, h: i32, c: Color) {
        if w <= 0 || h <= 0 {
            return;
        }
        self.draw_line(x, y, x + w - 1, y, c);
        self.draw_line(x, y + h - 1, x + w - 1, y + h - 1, c);
        self.draw_line(x, y, x, y + h - 1, c);
        self.draw_line(x + w - 1, y, x + w - 1, y + h - 1, c);
    }

    pub fn draw_line(&mut self, mut x0: i32, mut y0: i32, x1: i32, y1: i32, c: Color) {
        let dx = (x1 - x0).abs();
        let sx = if x0 < x1 { 1 } else { -1 };
        let dy = -(y1 - y0).abs();
        let sy = if y0 < y1 { 1 } else { -1 };
        let mut err = dx + dy;

        loop {
            self.put_pixel(x0, y0, c);
            if x0 == x1 && y0 == y1 {
                break;
            }
            let e2 = 2 * err;
            if e2 >= dy {
                err += dy;
                x0 += sx;
            }
            if e2 <= dx {
                err += dx;
                y0 += sy;
            }
        }
    }

    pub fn draw_circle(&mut self, cx: i32, cy: i32, radius: i32, c: Color) {
        if radius < 0 {
            return;
        }
        let mut x = radius;
        let mut y = 0;
        let mut err = 0;

        while x >= y {
            self.put_pixel(cx + x, cy + y, c);
            self.put_pixel(cx + y, cy + x, c);
            self.put_pixel(cx - y, cy + x, c);
            self.put_pixel(cx - x, cy + y, c);
            self.put_pixel(cx - x, cy - y, c);
            self.put_pixel(cx - y, cy - x, c);
            self.put_pixel(cx + y, cy - x, c);
            self.put_pixel(cx + x, cy - y, c);

            if err <= 0 {
                y += 1;
                err += 2 * y + 1;
            }
            if err > 0 {
                x -= 1;
                err -= 2 * x + 1;
            }
        }
    }

    pub fn fill_circle(&mut self, cx: i32, cy: i32, radius: i32, c: Color) {
        if radius < 0 {
            return;
        }
        let r2 = radius * radius;
        for dy in -radius..=radius {
            let mut dx_max = 0;
            while (dx_max + 1) * (dx_max + 1) + dy * dy <= r2 {
                dx_max += 1;
            }
            self.draw_line(cx - dx_max, cy + dy, cx + dx_max, cy + dy, c);
        }
    }

    pub fn draw_char(&mut self, x: i32, y: i32, ch: char, fg: Color, bg: Option<Color>) {
        let glyph = get_glyph(ch);
        for row in 0..8 {
            let byte = glyph[row];
            for col in 0..8 {
                if (byte & (0x80 >> col)) != 0 {
                    self.put_pixel(x + col as i32, y + row as i32, fg);
                } else if let Some(bg_col) = bg {
                    if bg_col.a > 0 {
                        self.put_pixel(x + col as i32, y + row as i32, bg_col);
                    }
                }
            }
        }
    }

    pub fn draw_text(&mut self, x: i32, y: i32, text: &str, fg: Color, bg: Option<Color>) {
        let mut cur_x = x;
        let mut cur_y = y;

        for ch in text.chars() {
            if ch == '\n' {
                cur_x = x;
                cur_y += 10;
            } else if ch == '\t' {
                cur_x += 32;
            } else {
                self.draw_char(cur_x, cur_y, ch, fg, bg);
                cur_x += 8;
            }
        }
    }

    /// Fast-blits the backbuffer to physical framebuffer in single burst
    pub fn swap_buffers(&mut self) {
        unsafe {
            ptr::copy_nonoverlapping(self.backbuffer.as_ptr(), self.mmap_ptr, self.smem_len);
        }
    }

    pub fn copy_to_backbuffer(&mut self, src: &[u8]) {
        let len = src.len().min(self.smem_len);
        self.backbuffer[..len].copy_from_slice(&src[..len]);
    }

    pub fn get_backbuffer(&self) -> &[u8] {
        &self.backbuffer
    }
}

impl Drop for Graphics {
    fn drop(&mut self) {
        // Restore VT text mode
        if let Some(fd) = self.tty_fd {
            unsafe {
                libc::ioctl(fd, KDSETMODE, KD_TEXT);
                libc::close(fd);
            }
        }

        // Unmap framebuffer memory
        if !self.mmap_ptr.is_null() && self.mmap_ptr != libc::MAP_FAILED as *mut u8 {
            unsafe {
                libc::munmap(self.mmap_ptr as *mut libc::c_void, self.smem_len);
            }
        }
    }
}
